### ISSUE TITLE
Semantic model does not resolve <see cref="" /> references to local methods and variables

### ISSUE THREAD

**Issue by HellBrick** [dotnet/roslyn#21215]
**Version Used**: 
VS 15.2 (26430.16)

**Steps to Reproduce**:
```csharp
namespace SeeCrefVersusLocalThings
{
	public class SomeClass
	{
		private readonly int _value = 42;

		public void OuterMethod()
		{
			/// <see cref="LocalMethod"/> is NOT resolved
			/// <see cref="PrivateMethod"/> is resolved
			/// <see cref="localVar"/> is NOT resolved
			/// <see cref="_value"/> is resolved

			int localVar = LocalMethod();

			int LocalMethod() => _value;
		}

		private int PrivateMethod() => _value;
	}
}
```

**Expected Behavior**:
All the functionality that targets the symbol and requires it to be resolved (Go to definition, Find all references, Rename, Ctrl+Shift+Up/Down, etc.) should work for all `<see cref=""/>` comments.

**Actual Behavior**:
It doesn't, at least not for `LocalMethod` and `localVar`.

**Comment by sharwell** [comment]
Documentation comments are not allowed to be placed on local variables or local methods. When they appear in this location, they are no different from normal comments from the perspective of the compiler.

It seems we have a bug in the presentation of these comments in the IDE, where they *look* like documentation comments and behave somewhat like documentation comments, even though they are not.

:bulb: You may have the documentation comment file turned off for your project, in which case the compiler treats *all* comments like regular comments and fails to report warnings when they are misused/invalid. If you enable documentation comment output as part of the build, your example above will produce several instances of CS1598 (XML comment is not placed on a valid language element).

**Comment by HellBrick** [comment]
Let me clarify my point: it doesn't really matter if the comments are placed *on* the local variables or local methods, it's about having proper tooling when the comments reference them. Even when not being used for generating public documentation, `<see cref=""/>` is extremely useful for referencing fields, methods and variables in internal comments describing the implementation details. Such comments can reside in a method's `<remarks>` section, but they can just as easily occur inline in the method or anywhere else. As of today, the compiler's support of `<see cref=""/>` already goes beyond the scope of the public documentation: the tooling has proper support for referencing the private members of the class, which doesn't really make sense for the public documentation, but helps with internal comments a lot. All I'm suggesting is extending this support to the local symbols.

**Comment by sharwell** [comment]
> Let me clarify my point: it doesn't really matter if the comments are placed on the local variables or local methods, it's about having proper tooling when the comments reference them.

Right, the bug here is that the IDE shows the comment in your example visually different from the following comment:

```csharp
// <see cref="LocalMethod"/> is NOT resolved
// <see cref="PrivateMethod"/> is resolved
// <see cref="localVar"/> is NOT resolved
// <see cref="_value"/> is resolved
```

By displaying things differently, it suggests that a feature like rename or finding references would work. However, the comments in your example have no semantic content according to the language. It's not that the reference isn't resolved, it's that the reference doesn't exist in the first place.

Note that I'm not saying this is the way the language should be - to the contrary, I've long believed that the limitations on the placement of semantic comments is a leading area for improvement of modern programming languages.

**Comment by HellBrick** [comment]
But the reference *does* currently exist for the private members =) Here's the rename UI, note the "include comments" being unchecked:
![Screenshot of the rename targeting a comment reference to a private member](https://i.imgur.com/fOFNkDB.png)

I believe it can (and should) work similarly for the local methods and variables.

And to avoid the confusion caused by the comments located in an "inappropriate place", let's adjust the sample this way:
```csharp
namespace SeeCrefVersusLocalThings
{
	public class SomeClass
	{
		private readonly int _value = 42;

		/// <remarks>
		/// <see cref="LocalMethod"/> is NOT resolved
		/// <see cref="PrivateMethod"/> is resolved
		/// <see cref="localVar"/> is NOT resolved
		/// <see cref="_value"/> is resolved
		/// </remarks>
		public void OuterMethod()
		{
			int localVar = LocalMethod();

			int LocalMethod() => _value;
		}

		private int PrivateMethod() => _value;
	}
}
```

When written this way, it's a bit less obvious how the compiler should resolve the local variables when the comment is written outside the scope where the local can be seen. On the other hand, the compiler can work this out for the private member references in the comments at the class level, so the similar logic can probably apply:

![The screenshot of referencing the private field from a class-level comment](https://i.imgur.com/CckZm9f.png)

**Comment by sharwell** [comment]
This might make it easier - here is the behavior I would expect, given the current language specification:

```xml
/// <remarks>
/// <see cref="LocalMethod"/> is NOT resolved
/// <see cref="PrivateMethod"/> is resolved
/// <see cref="localVar"/> is NOT resolved
/// <see cref="_value"/> is resolved
/// </remarks>
public void OuterMethod()
{
	/// <remarks>
	/// <see cref="LocalMethod"/> is NOT resolved
	/// <see cref="PrivateMethod"/> is NOT resolved
	/// <see cref="localVar"/> is NOT resolved
	/// <see cref="_value"/> is NOT resolved
	/// </remarks>
	int localVar = LocalMethod();

	int LocalMethod() => _value;
}
```

There are 3 bugs currently:

1. The syntax highlighting for the `///` comment which appears before `LocalMethod` is incorrect. It is being treated as a documentation comment, but in reality it is *not* a documentation comment.
2. Inside the `///` comment which is being treated incorrectly, code which resembles an XML reference to `PrivateMethod` is being "resolved". This is incorrect because the text is not a reference and is not part of an XML documentation comment.
3. For the same reason as the preceding item, the text `_value` is incorrectly getting "resolved".

**Comment by sharwell** [comment]
:memo: Also note I don't actually think we need to make any change to fix the issues I mentioned. I would very much prefer instead to update the language to support some of these new scenarios.

**Comment by HellBrick** [comment]
If we consider just the correctly placed comment for a second, would adding support for its `<see cref>`s to reference local symbols, just like they can reference private symbols today, be considered a simple compiler bugfix/improvement? Or should an issue be raised in the https://github.com/dotnet/csharplang repo to discuss it there first?

I keep drawing this parallel between referencing private methods and referencing local methods because neither makes sense for the documentation comments, while both are useful for internal comments. So from my point of view, either both of them should be supported by introducing support for referencing local methods, or none of them should not be supported by dropping support for referencing private methods.

**Comment by sharwell** [comment]
> Or should an issue be raised in the https://github.com/dotnet/csharplang repo to discuss it there first?

For locals (variables and functions), it's likely the change would need to start there.

**Comment by agocke** [comment]
@jaredpar I agree with @sharwell here. Doc comments are not specced to work in blocks or on locals. This looks like a feature request to me.

**Comment by jaredpar** [comment]
That was my intuition as well. Moved to feature request.
